### PR TITLE
Active-Failover with Velocystream

### PR DESCRIPTION
**Pull request by graetzer**
This branch will correspond to this pull request: https://github.com/arangodb/arangodb/pull/4715

### CODE REVIEW

File: replication.go [hunk 1]
```diff
@@ -24,12 +24,24 @@ package driver
...
 type Replication interface {
+	// CreateBatch creates a "batch" to prevent WAL file removal and to take a snapshot
+	CreateBatch(ctx context.Context, serverID int64, db Database, ttl time.Duration) (BatchMetadata, error)
+	// DeleteBatch deletes an existing dump batch
+	DeleteBatch(ctx context.Context, db Database, batchID string) error
 	// Get the inventory of the server containing all collections (with entire details) of a database.
 	// When this function is called on a coordinator is a cluster, an ID of a DBServer must be provided
 	// using a context that is prepare with `WithDBServerID`.
-	DatabaseInventory(ctx context.Context, db Database) (DatabaseInventory, error)
+	DatabaseInventory(ctx context.Context, db Database, batchID string) (DatabaseInventory, error)
```
Comment (ewoutp) [inline replication.go]: This is a breaking change in the API!
If a batchID has to be brought in (it is certainly not needed in all cases), use a context variable (e.g. `WithBatchID`)
See context.go

Comment (graetzer) [reply replication.go]: The batchID is needed to get the inventory when we use rocksdb. So this API could not have worked for anybody (using rocksdb), I do not see it as a huge issue to add a new method.

Maybe that line of thought is wrong ?

Comment (ewoutp) [reply replication.go]: In Arangosync the DatabaseInventory is already used. It calls a dbserver directly.

Comment (ewoutp) [reply replication.go]: What do you mean add a new method?

Comment (graetzer) [reply replication.go]: Sorry I meant write "add a new argument". Arangosync uses `DatabaseInventory` from the `Cluster` interface, no?

Comment (ewoutp) [reply replication.go]: Again, adding a new argument is breaking is API. Not good.
Please use the context way.

File: replication_impl.go [hunk 2]
```diff
@@ -25,8 +25,55 @@ package driver
...
+}
+
+// CreateBatch creates a "batch" to prevent WAL file removal and to take a snapshot
+func (c *client) CreateBatch(ctx context.Context, serverID int64, db Database) (BatchMetadata, error) {
+	req, err := c.conn.NewRequest("POST", path.Join("_db", db.Name(), "_api/replication/batch"))
+	if err != nil {
+		return BatchMetadata{}, WithStack(err)
+	}
+	req = req.SetQuery("serverId", strconv.FormatInt(serverID, 10))
+	params := batch_params{ttl: 60.0} // just use a default ttl value
```
Comment (ewoutp) [inline replication_impl.go]: First: The default TTL should be specified in a `const`.
Second: I suggest to make the TTL configurable with a context function `WithBatchTTL`

File: test/client_test.go [hunk 3]
```diff
@@ -195,7 +195,7 @@ func waitUntilServerAvailable(ctx context.Context, c driver.Client, t testEnv) b
 				return
 			} else {
 				cancel()
-				//t.Logf("Version failed: %s %#v", describe(err), err)
+				t.Logf("Version failed: %s %#v", describe(err), err)
```
Comment (ewoutp) [inline test/client_test.go]: Restore this to original before the PR gets merged

File: test/cluster_test.go [hunk 4]
```diff
@@ -38,6 +38,9 @@ func TestClusterHealth(t *testing.T) {
 	if driver.IsPreconditionFailed(err) {
 		t.Skip("Not a cluster")
 	} else {
+		if err != nil {
```
Comment (ewoutp) [inline test/cluster_test.go]: Change to 
```
} else if err != nil {
  t.Fatalf("Cluster failed: %s, describe(err))
} else {
  ...
```

File: test/replication_test.go [hunk 5]
```diff
@@ -45,7 +45,24 @@ func TestReplicationDatabaseInventory(t *testing.T) {
...
 			t.Fatalf("Failed to open _system database: %s", describe(err))
 		}
-		inv, err := rep.DatabaseInventory(ctx, db)
+
+		version, err := c.Version(nil)
+		if err != nil {
+			t.Fatalf("Version failed: %s", describe(err))
+		}
+
+		ctx_inv := ctx
```
Comment (ewoutp) [inline test/replication_test.go]: Change underscore to camelcase

File: test/user_auth_test.go [hunk 6]
```diff
@@ -258,6 +252,11 @@ func TestGrantUserDefaultDatabase(t *testing.T) {
 		t.Fatalf("Expected success, got %s", describe(err))
 	}
 
+	// Grant read-only access to default database
```
Comment (ewoutp) [inline test/user_auth_test.go]: Why moving this code code?
This changes the expected behavior, which may be a problem

File: test/user_auth_test.go [hunk 7]
```diff
@@ -390,21 +381,42 @@ func TestGrantUserCollection(t *testing.T) {
 	if _, err := authCol.ReadDocument(nil, meta1.Key, &doc); !driver.IsForbidden(err) {
 		t.Errorf("Expected failure, got %s", describe(err))
 	}
-
 	// Now remove explicit collection access
 	if err := u.RemoveCollectionAccess(nil, col); err != nil {
 		t.Fatalf("RemoveCollectionAccess failed: %s", describe(err))
 	}
+	expected := driver.GrantNone
```
Comment (ewoutp) [inline test/user_auth_test.go]: What is this?
After version 3.3.4 we have an (behavioural) API change?

Comment (graetzer) [reply test/user_auth_test.go]: Yes 3.3.4 changes some behaviour, it mentioned on the CHANGELOG.

File: test/user_auth_test.go (hunk 6 above)
Comment (graetzer) [reply test/user_auth_test.go]: Yes 3.3.4 changes some behaviour, it mentioned on the CHANGELOG. This was decided to make the use of the LDAP backend easier.

File: replication.go [hunk 8]
```diff
@@ -24,10 +24,31 @@ package driver
...
+	// Extend the lifetime of an existing batch on the server
+	Extend(ctx context.Context, ttl time.Duration) error
+	// DeleteBatch deletes an existing batch on the server
+	Delete(ctx context.Context) error
+}
+
 // Replication provides access to replication related operations.
 type Replication interface {
+	// CreateBatch creates a "batch" to prevent removal of state required for replication
+	CreateBatch(ctx context.Context, serverID int64, db Database, ttl time.Duration) (Batch, error)
```
Comment (ewoutp) [inline replication.go]: Have a light preference for `ctx, db, serverID, ttl`

File: replication_impl.go [hunk 9]
```diff
@@ -47,3 +93,61 @@ func (c *client) DatabaseInventory(ctx context.Context, db Database) (DatabaseIn
 	}
 	return result, nil
 }
+
+func (b batchMetadata) BatchID() string {
```
Comment (ewoutp) [inline replication_impl.go]: Comment required

File: replication_impl.go [hunk 10]
```diff
@@ -47,3 +93,61 @@ func (c *client) DatabaseInventory(ctx context.Context, db Database) (DatabaseIn
...
+	return b.ID
+}
+
+func (b batchMetadata) Tick() Tick {
```
Comment (ewoutp) [inline replication_impl.go]: Comment required

File: replication_impl.go [hunk 11]
```diff
@@ -47,3 +93,61 @@ func (c *client) DatabaseInventory(ctx context.Context, db Database) (DatabaseIn
...
+	return b.LastTick
+}
+
+// Extend the lifetime of an existing batch on the server
+func (b batchMetadata) Extend(ctx context.Context, ttl time.Duration) error {
+	if b.closed != 0 {
```
Comment (ewoutp) [inline replication_impl.go]: Probably you want to use `CompareAndSwap` here as well

File: test/user_auth_test.go (hunk 6 above)
Comment (ewoutp) [reply test/user_auth_test.go]: Ok. Please make sure this still works on single, resilientsingle & cluster in all permutations.

File: replication_impl.go [hunk 12]
```diff
@@ -47,3 +93,63 @@ func (c *client) DatabaseInventory(ctx context.Context, db Database) (DatabaseIn
...
+
+// Last tick reported by this batch
+func (b batchMetadata) Tick() Tick {
+	return b.LastTick
+}
+
+// Extend the lifetime of an existing batch on the server
+func (b batchMetadata) Extend(ctx context.Context, ttl time.Duration) error {
+	if !atomic.CompareAndSwapInt32(&b.closed, 0, 0) {
+		return nil
```
Comment (ewoutp) [inline replication_impl.go]: Seems this is not good. 
When closed, it should return an error, not nil

File: context.go [hunk 13]
```diff
@@ -206,6 +207,11 @@ func WithDBServerID(parent context.Context, id string) context.Context {
 	return context.WithValue(contextOrBackground(parent), keyDBServerID, id)
 }
 
+// WithBatchID is used to configure a context that includes an ID of a Batch.
```
[the remark this answers is not in the document]
Comment (ewoutp) [reply context.go]: Comment is still relevant

File: replication_impl.go [hunk 14]
```diff
@@ -47,3 +93,63 @@ func (c *client) DatabaseInventory(ctx context.Context, db Database) (DatabaseIn
...
+	if err := resp.CheckStatus(204); err != nil {
+		return WithStack(err)
+	}
+	return nil
+}
+
+// DeleteBatch deletes an existing dump batch
+func (b *batchMetadata) Delete(ctx context.Context) error {
+	if !atomic.CompareAndSwapInt32(&b.closed, 0, 1) {
+		return nil
```
Comment (ewoutp) [inline replication_impl.go]: Return an error

File: test/replication_test.go [hunk 15]
```diff
@@ -45,7 +46,24 @@ func TestReplicationDatabaseInventory(t *testing.T) {
...
+
+		version, err := c.Version(nil)
+		if err != nil {
+			t.Fatalf("Version failed: %s", describe(err))
+		}
+
+		ctx2 := ctx
+		if version.Version.CompareTo("3.2") >= 0 {
+			// RocksDB requires batchID
+			batch, err := rep.CreateBatch(ctx, db, 1337, time.Duration(60.0))
```
Comment (ewoutp) [inline test/replication_test.go]: Please add a comment explaining where the `1337` comes from.
Or change to:
```
serverID := 1337 // Random test value
batch, err := rep.CreateBatch(ctx, db, 1337, time.Second * 60)
```

Also not the change in time. I doubt 60ns is the intended value

File: replication_impl.go [hunk 16]
```diff
@@ -47,3 +93,63 @@ func (c *client) DatabaseInventory(ctx context.Context, db Database) (DatabaseIn
 	}
 	return result, nil
 }
+
+// id of this batch
```
Comment (ewoutp) [inline replication_impl.go]: Comment of go member must start with the name of the member.
So comment should be:
```
// BatchID returns the ID of this batch
```

File: replication_impl.go [hunk 17]
```diff
@@ -47,3 +93,63 @@ func (c *client) DatabaseInventory(ctx context.Context, db Database) (DatabaseIn
...
+func (b batchMetadata) BatchID() string {
+	return b.ID
+}
+
+// Last tick reported by this batch
```
Comment (ewoutp) [inline replication_impl.go]: Comment must start with `Tick` (see above)

File: replication.go [hunk 18]
```diff
@@ -24,10 +24,31 @@ package driver
...
+type Tick string
+
+// Batch represents state on the server used during
+// certain replication operations to keep state required
+// by the client (such as Write-Ahead Log, inventory and data-files)
+type Batch interface {
+	// id of this batch
+	BatchID() string
+	// Last tick reported by this batch
+	Tick() Tick
```
Comment (ewoutp) [inline replication.go]: If it returns the last tick, then I would prefer this function to be called `LastTick`

**Pull request merged**